File: sdk/src/main/java/ru/usedesk/sdk/ui/knowledgebase/FragmentDataView.java
package ru.usedesk.sdk.ui.knowledgebase;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import ru.usedesk.sdk.R;

public abstract class FragmentDataView<V, T extends DataViewModel<V>> extends FragmentView<T> {

    private final int layoutId;

    private TextView textViewMessage;

    private View container;

    public FragmentDataView(int layoutId) {
        this.layoutId = layoutId;
    }

    protected abstract void setDataView(V data);

    protected abstract ViewModelFactory<T> getViewModelFactory();

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(layoutId, container, false);

        onView(view);

        initViewModel(getViewModelFactory());

        getViewModel().getLiveData()
                .observe(this, this::onData);

        return view;
    }

    protected void onView(@NonNull View view) {
        textViewMessage = view.findViewById(R.id.tv_message);
        container = view.findViewById(R.id.container);
    }

    protected void onData(DataOrMessage<V> dataOrMessage) {
        switch (dataOrMessage.getMessage()) {
            case LOADING:
                onMessage(R.string.loading_title);
                break;
            case ERROR:
                onMessage(R.string.loading_error);
                break;
            default:
                onData(dataOrMessage.getData());
                break;
        }
    }

    protected void onData(V data) {
        setDataView(data);

        textViewMessage.setVisibility(View.GONE);
        container.setVisibility(View.VISIBLE);
    }

    private void onMessage(int resourceId) {
        textViewMessage.setText(resourceId);

        textViewMessage.setVisibility(View.VISIBLE);
        container.setVisibility(View.GONE);
    }
}

File: knowledgebase-gui/src/main/java/ru/usedesk/knowledgebase_gui/common/FragmentDataView.kt
package ru.usedesk.knowledgebase_gui.common

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.lifecycle.LiveData
import ru.usedesk.common_gui.UsedeskBinding
import ru.usedesk.common_gui.UsedeskFragment
import ru.usedesk.common_gui.inflateItem
import ru.usedesk.knowledgebase_gui.R
import ru.usedesk.knowledgebase_gui.entity.DataOrMessage

abstract class FragmentDataView<DATA, BINDING : UsedeskBinding>(
        private val layoutId: Int
) : UsedeskFragment() {

    private lateinit var textViewMessage: TextView
    private lateinit var binding: BINDING
    private lateinit var container: View

    protected abstract fun setDataView(data: DATA)

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View {
        binding = inflateItem(inflater, container, layoutId) {
            createBinding(it)
        }

        onView(binding.rootView)

        init()

        getLiveData().observe(viewLifecycleOwner, {
            this.onData(it)
        })

        return binding.rootView
    }

    abstract fun createBinding(rootView: View): BINDING

    open fun init() {}

    abstract fun getLiveData(): LiveData<DataOrMessage<DATA>>

    protected open fun onView(view: View) {
        textViewMessage = view.findViewById(R.id.tv_message)
        container = view.findViewById(R.id.container)
    }

    private fun onData(dataOrMessage: DataOrMessage<DATA>) {
        when (dataOrMessage.message) {
            DataOrMessage.Message.LOADING -> onMessage(R.string.loading_title)
            DataOrMessage.Message.ERROR -> onMessage(R.string.loading_error)
            else -> onData(dataOrMessage.data)
        }
    }

    private fun onData(data: DATA?) {
        if (data != null) {
            setDataView(data)
            textViewMessage.visibility = View.GONE
            container.visibility = View.VISIBLE
        }
    }

    private fun onMessage(resourceId: Int) {
        textViewMessage.setText(resourceId)
        textViewMessage.visibility = View.VISIBLE
        container.visibility = View.GONE
    }
}